tracim/tracim/lib/email_fetcher.py

```diff
@@ -0,0 +1,323 @@
+
+    def get_key(self) -> typing.Optional[str]:
+
+        """
+        key is the string contain in some mail header we need to retrieve.
+        First try checking special header, them check 'to' header
+        and finally check first(oldest) mail-id of 'references' header
+        """
+        first_ref = self.get_first_ref()
+        to_address = self.get_to_address()
+        special_key = self.get_special_key()
+
+        if special_key:
+            return special_key
+        if to_address:
+            return DecodedMail.find_key_from_mail_address(to_address)
+        if first_ref:
+            return DecodedMail.find_key_from_mail_address(first_ref)
+
+    @classmethod
+    def find_key_from_mail_address(cls, mail_address: str) \
+            -> typing.Optional[str]:
+        """ Parse mail_adress-like string
+        to retrieve key.
+
+        :param mail_address: user+key@something like string
+        :return: key
+        """
+        username = mail_address.split('@')[0]
+        username_data = username.split('+')
+        if len(username_data) == 2:
+            return username_data[1]
+        return None
+
+
+class MailFetcher(object):
+
+    def __init__(self,
+                 host: str,
+                 port: str,
+                 user: str,
+                 password: str,
+                 use_ssl: bool,
+                 folder: str,
+                 delay: int,
+                 endpoint: str,
+                 token: str) \
+            -> None:
+        """
+        Fetch mail from a mailbox folder through IMAP and add their content to
+        Tracim through http according to mail Headers.
+        Fetch is regular.
+        :param host: imap server hostname
+        :param port: imap connection port
+        :param user: user login of mailbox
+        :param password: user password of mailbox
+        :param use_ssl: use imap over ssl connection
+        :param folder: mail folder where new mail are fetched
+        :param delay: seconds to wait before fetching new mail again
+        :param endpoint: tracim http endpoint where decoded mail are send.
+        :param token: token to authenticate http connexion
+        """
+        self._connection = None
+        self.host = host
+        self.port = port
+        self.user = user
+        self.password = password
+        self.use_ssl = use_ssl
+        self.folder = folder
+        self.delay = delay
+        self.endpoint = endpoint
+        self.token = token
+
+        self._is_active = True
+
+    def run(self) -> None:
+        while self._is_active:
+            time.sleep(self.delay)
+            try:
+                self._connect()
+                mails = self._fetch()
+                # TODO - G.M -  2017-11-22 retry sending unsended mail
+                # These mails are return by _notify_tracim, flag them with "unseen"
+                # or store them until new _notify_tracim call
+                self._notify_tracim(mails)
```

Comment:
Actually, I just saw that notifying tracim and getting "cleaned mail" are done in the same task, which is not a good way to work (it's difficult to understand where is the part of the algorithm in charge of "purging emails" from trash stuff.

Instead of

```
self._connect()
mails = self._fetch()
self._notify_tracim(mails)
self._disconnect()
```

You should have something like this:

```
self._connect()
messages = self._fetch()  # returns Message objects, not DecodedMails
cleaned_mails = [DecodedMail(msg) for msg in messages]
self._notify_tracim(cleaned_mails)
self._disconnect()
```

Why ? Writing the code this way allow to insert other operations without to modify _fetch() (which is in charge of fetching, not in charge of cleaning emails)

Could you add a ticket for a future refactoring? (or fix this before we merge)